pkg/chartsync/git.go

```diff
@@ -0,0 +1,337 @@
+
+	return &GitChartSync{
+		logger:       logger,
+		config:       cfg,
+
+		lister:       lister,
+
+		mirrors:      git.NewMirrors(),
+		sources:      make(map[string]sourceRef),
+
+		releaseQueue: queue,
+	}
+}
+
+// Run starts the mirroring of git repositories, and processes mirror
+// changes on signal, scheduling a release for a `HelmRelease` resource
+// when the update is relevant to the release.
+func (c *GitChartSync) Run(stopCh <-chan struct{}, errCh chan error, wg *sync.WaitGroup) {
+	c.logger.Log("info", "starting sync of git chart sources")
+
+	wg.Add(1)
+	go func() {
+		defer func() {
+			c.mirrors.StopAllAndWait()
+			wg.Done()
+		}()
+
+		for {
+			select {
+			case changed := <-c.mirrors.Changes():
+				for mirrorName := range changed {
+					repo, ok := c.mirrors.Get(mirrorName)
+
+					hrs, err := c.helmReleasesForMirror(mirrorName)
+					if err != nil {
+						c.logger.Log("error", ErrReleasesForMirror.Error(), "mirror", mirrorName, "err", err)
+						continue
+					}
+
+					// We received a signal from a no longer existing
+					// mirror.
+					if !ok {
+						if len(hrs) == 0 {
+							// If there are no references to it either,
+							// just continue with the next mirror...
+							continue
+						}
+
+						c.logger.Log("warning", ErrNoMirror.Error(), "mirror", mirrorName)
+						for _, hr := range hrs {
+							c.maybeMirror(mirrorName, hr.Spec.GitChartSource.GitURL)
+						}
+						// Wait for the signal from the newly requested mirror...
+						continue
+					}
+
+					c.processChangedMirror(mirrorName, repo, hrs)
+				}
+			case <-stopCh:
+				c.logger.Log("info", "stopping sync of git chart sources")
+				return
+			}
+		}
+	}()
+}
+
+// ChartDir returns a newly exported `GitChartDir` for the given `v1.HelmRelease`
+// at the latest HEAD, or an error.
+func (c *GitChartSync) ChartDir(hr *v1.HelmRelease) (GitChartDir, error) {
+	mirror := mirrorName(hr)
+	repo, ok := c.mirrors.Get(mirror)
+	if !ok {
+		// We did not find a mirror; request one, return, and wait for
+		// signal.
+		c.maybeMirror(mirror, hr.Spec.GitURL)
+		return GitChartDir{}, ChartNotReadyError{ErrNoMirror}
+	}
+
+	s, ok, err := c.sync(hr, mirror, repo)
+	if err != nil {
+		return GitChartDir{}, err
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), c.config.GitTimeout)
+	defer cancel()
+	export, err := repo.Export(ctx, s.head)
+	if err != nil {
+		return GitChartDir{}, ChartUnavailableError{err}
+	}
+
+	return GitChartDir{export, s.head}, nil
+}
+
+// Delete cleans up the source reference for the given `v1.HelmRelease`,
+// this includes the mirror if there is no reference to it from sources.
+// It returns a boolean indicating a successful removal (`true` if so,
+// `false` otherwise).
+func (c *GitChartSync) Delete(hr *v1.HelmRelease) bool {
+	c.sourcesMu.Lock()
+	defer c.sourcesMu.Unlock()
+
+	// Attempt to get the source from store.
+	source, ok := c.sources[hr.ResourceID().String()]
+	if ok {
+		// Remove the in store source.
```

Comment:
Remove the ... release?

Reply:
We are not deleting the release here but our indexation / bookkeeping of the git source. I think the comment does not add much now that the code has been reduced.

Reply:
Ah, you meant the _in-store_ source (sorry, I should had understood)

Reply:
Yes, I already thought the missing hyphen was the actual issue.